File: src/significant_climbs.py
"""Shared climb significance rules for analysis export and Companion bundles."""

from __future__ import annotations

from typing import Any, Protocol


class _ClimbMetrics(Protocol):
    length_km: float
    elevation_gain_m: float
    avg_gradient_pct: float


def _max_gradient_pct(climb: Any) -> float:
    values: list[float] = []
    for key in (
        "max_50_m_pct",
        "max_100_m_pct",
        "max_250_m_pct",
        "max_500_m_pct",
        "max_1000_m_pct",
    ):
        value = getattr(climb, key, None)
        if value is None and isinstance(climb, dict):
            value = climb.get(key)
        if value is not None:
            values.append(float(value))
    return max(values) if values else 0.0


def _length_km(climb: Any) -> float:
    if isinstance(climb, dict):
        return float(climb.get("length_km") or climb.get("lengthKm") or 0)
    return float(getattr(climb, "length_km", 0) or getattr(climb, "lengthKm", 0) or 0)


def _gain_m(climb: Any) -> float:
    if isinstance(climb, dict):
        return float(climb.get("elevation_gain_m") or climb.get("elevationGainM") or 0)
    return float(
        getattr(climb, "elevation_gain_m", 0) or getattr(climb, "elevationGainM", 0) or 0
    )


def _avg_gradient_pct(climb: Any) -> float:
    if isinstance(climb, dict):
        return float(climb.get("avg_gradient_pct") or climb.get("avgGradientPct") or 0)
    return float(
        getattr(climb, "avg_gradient_pct", 0) or getattr(climb, "avgGradientPct", 0) or 0
    )


def is_significant_climb(climb: Any) -> bool:
    """Return True when a climb is meaningful to riders, not a short GPX bump."""
    gain_m = _gain_m(climb)
    length_km = _length_km(climb)
    avg_gradient_pct = _avg_gradient_pct(climb)
    max_gradient_pct = _max_gradient_pct(climb)

    if gain_m < 100:
        return False
    if length_km < 2 and gain_m < 150:
        return False
    if gain_m >= 200:
        return True
    if length_km >= 5 and gain_m >= 150 and (
        avg_gradient_pct >= 2.5 or max_gradient_pct >= 6
    ):
        return True
    if length_km >= 3 and gain_m >= 120 and max_gradient_pct >= 5:
        return True
    return False


def significant_climbs(climbs: list[Any]) -> list[Any]:
    return [climb for climb in climbs if is_significant_climb(climb)]

File: shared/race/significantClimbs.ts
/** Shared climb significance rules for Desktop, Companion, and analysis export. */

export interface SignificantClimbMetrics {
  length_km?: number;
  lengthKm?: number;
  elevation_gain_m?: number;
  elevationGainM?: number;
  avg_gradient_pct?: number;
  avgGradientPct?: number;
  max_50_m_pct?: number | null;
  max_100_m_pct?: number | null;
  max_250_m_pct?: number | null;
  max_500_m_pct?: number | null;
  max_1000_m_pct?: number | null;
  max50mPct?: number | null;
  max100mPct?: number | null;
  max250mPct?: number | null;
  max500mPct?: number | null;
  max1000mPct?: number | null;
}

function climbLengthKm(climb: SignificantClimbMetrics): number {
  return climb.length_km ?? climb.lengthKm ?? 0;
}

function climbGainM(climb: SignificantClimbMetrics): number {
  return climb.elevation_gain_m ?? climb.elevationGainM ?? 0;
}

function climbAvgGradientPct(climb: SignificantClimbMetrics): number {
  return climb.avg_gradient_pct ?? climb.avgGradientPct ?? 0;
}

function climbMaxGradientPct(climb: SignificantClimbMetrics): number {
  const values = [
    climb.max_50_m_pct ?? climb.max50mPct,
    climb.max_100_m_pct ?? climb.max100mPct,
    climb.max_250_m_pct ?? climb.max250mPct,
    climb.max_500_m_pct ?? climb.max500mPct,
    climb.max_1000_m_pct ?? climb.max1000mPct,
  ].filter((value): value is number => value != null && Number.isFinite(value));
  return values.length > 0 ? Math.max(...values) : 0;
}

/**
 * A climb is significant when total gain and length matter to riders,
 * not when a short GPX bump barely clears minimum detector thresholds.
 */
export function isSignificantClimb(climb: SignificantClimbMetrics): boolean {
  const gainM = climbGainM(climb);
  const lengthKm = climbLengthKm(climb);
  const avgGradientPct = climbAvgGradientPct(climb);
  const maxGradientPct = climbMaxGradientPct(climb);

  if (gainM < 100) {
    return false;
  }
  if (lengthKm < 2 && gainM < 150) {
    return false;
  }
  if (gainM >= 200) {
    return true;
  }
  if (lengthKm >= 5 && gainM >= 150 && (avgGradientPct >= 2.5 || maxGradientPct >= 6)) {
    return true;
  }
  if (lengthKm >= 3 && gainM >= 120 && maxGradientPct >= 5) {
    return true;
  }
  return false;
}

export function significantClimbs<T extends SignificantClimbMetrics>(climbs: T[]): T[] {
  return climbs.filter(isSignificantClimb);
}
